lib/connectors/memory.js

```diff
@@ -250,6 +250,46 @@ Memory.prototype.updateOrCreate = function(model, data, options, callback) {
     }
   });
 };
+Memory.prototype.patchOrCreateWithWhere =
+  Memory.prototype.upsertWithWhere = function(model, where, data, options, callback) {
+    var self = this;
+    var element = Object.keys(where)[0];
+    console.log (element);
+    var filter = { where: where };
+    console.log(filter);
+    console.log(model);
+    if (!data.hasOwnProperty(element)) {
+      data[element] = where[element];
+    }
+    var primaryKey = this.idName(model);
+    self.all(model, filter, options, function(err, instances) {
+      if (err) callback(err);
+      else {
+        var modelsLength = instances.length;
+        if (modelsLength === 0 && data.hasOwnProperty(primaryKey)) {
+          self.create(model, data, options, function(err, id) {
+            self.setIdValue(model, data, id);
+            callback(err, data, { isNewInstance: true });
+          });
+        } else if (modelsLength == 0 && data.hasOwnProperty(primaryKey) == false) {
+          var error = new Error('Error message: No matching instance found.Found error ' +
+            'while trying to create a new instance. Primary key is not provided ' +
+            'in the payload');
+          callback(error);
+        } else if (modelsLength === 1) {
+          var primaryKeyValue = instances[0][primaryKey];
+          return self.updateAttributes(model, primaryKeyValue, data, options, callback);
+        } else {
+          process.nextTick(function() {
+            var error = new Error('There are multiple instances found.' +
+              'Upsert Operation will not be performed');
```

Comment:
It makes perfect sense to me to fail `upsertWithWhere` when there are multiple records found.

See https://github.com/strongloop/loopback-datasource-juggler/pull/1001#issuecomment-232441096
> > 1) why we don't update multiple instances rather return error if multiple instances are found.
> This was discussed as part of the requirement we had and we plan to restrict this only for single instance update. I had a look at how upsert with foreign key is performed in some of the popularly used endpoints and they follow the similar approach.